### ISSUE TITLE
Connection Return Race Conditions [JIRA: CLIENTS-798]

### ISSUE THREAD

**Issue by alexmoore** [basho/riak-java-client#603]
There's a race condition between setting the `done` condition in `FutureOperation` and returning a connection in `RiakNode`.  This has only been seen in Linux so far. 

Ideally, we should return a connection before an operation is "done", so that it's available for the next operation on the user thread.  

When we call [`latch.countdown()`](https://github.com/basho/riak-java-client/blob/develop/src/main/java/com/basho/riak/client/core/FutureOperation.java#L169) from the response handler on the netty thread, this signals that the operation is complete.  This in turn unblocks anybody waiting on the application thread in the [`get()`](https://github.com/basho/riak-java-client/blob/develop/src/main/java/com/basho/riak/client/core/FutureOperation.java#L253-L257) method.  
After counting down the latch, we then [return the connection](https://github.com/basho/riak-java-client/blob/develop/src/main/java/com/basho/riak/client/core/RiakNode.java#L864-L869).  

The issue is that the application thread can race ahead and execute + retry the next operation 3 times if there is a pause on the netty thread after setting the result but before returning the connection.  This can result in `NoNodesAvailable` exceptions in applications where all connections are mostly in use. 

[RiakNode::onSuccess](https://github.com/basho/riak-java-client/blob/develop/src/main/java/com/basho/riak/client/core/RiakNode.java#L853-L872) [FutureOperation::setResponse](https://github.com/basho/riak-java-client/blob/develop/src/main/java/com/basho/riak/client/core/FutureOperation.java#L155-L172).

**Comment by kodesohard** [comment]
Is this issue fixed in latest java client 2.0.6? If not, do we know which version this fix will be in?

**Comment by alexmoore** [comment]
@kodesohard Yep! It was fixed in 2.0.6, but this this ticket ended up being a duplicate of #523, so I'll close it now.

Thanks!
